### PR TITLE
Tests that fail based on ROL segments occurring in multiple places in the same parent group

### PR DESCRIPTION
**Pull request by blendersfun**
The ADT_A03 message definition allows ROL segments to occur in multiple distinct locations. If they occur after the PID, PD1 section, then they are patient-level personnel. If they occur after the PV1, PV2 section they are visit-level personnel. This seems to clash a bit with the premise of the library. If you use segment codes as the output JSON keys, then you get a naming collision in this situation.

It seems to me that this could perhaps be fixed using some kind of suffix to disambiguate. For example, "ROL--PID", vs "ROL--PV1" could be used to distinguish "patient" vs "visit" positions in the parent ADT_A03 group. However, that would constitute a breaking change in the output json. Also, it would involve some potentially extensive changes to the schemas to add a "suffix" or "disambiguation" param as needed, and perhaps a schema validation test to assert that no ambiguous segments in parent groups exist without disambiguation. These are just the rough ideas I've had so far.

An alternative idea for disambiguation that would not require changes to the schemas would be to just add an auto-generated numeric index. In other words the first ROL in the A03 message would become "ROL1", the second would be "ROL2", etc. This could be generated on the fly when the group/segment info is being indexed.

We are investigating using this library for hl7 processing, but we currently depend on the ability to parse these two different types of ROL segments in A03 messages, and thus will need to come up with a fix before we can move forward. I have written an hl7-to-json parser before and have code reviewed the parser code in some detail. If I have time I will try to put together a proposal for a fix to this, however, I'd be interested in hearing feedback from the authors about the issue as well, particularly about the design aspect.

As an aside, I reviewed the contribution guidelines but there is a broken link. The "code of conduct" link directs me to https://js.foundation/community/code-of-conduct, which redirects to https://openjsf.org/. I'm not sure from there where the intended code of conduct document is.

### CODE REVIEW

File: lib/generator/message.js [hunk 1]
```diff
@@ -36,13 +36,14 @@ module.exports = function writeMessage (json, delimiters) {
       element = messageDef[groupName].elements[i];
 
       if (element.segment) {
-        if (groupJSON[element.segment]) {
+        const jsonKey = element.jsonKey || element.segment
+        if (groupJSON[jsonKey]) {
 
-          if (parseInt(element.maxOccurs === 1) || !(groupJSON[element.segment] instanceof Array)) {
-            ret += self.writeSegment(groupJSON[element.segment], element.segment, delimiters);
+          if (parseInt(element.maxOccurs === 1) || !(groupJSON[jsonKey] instanceof Array)) {
```
Comment (sdhakal-xealth) [inline lib/generator/message.js:42]: @x-aaron-moore fyi, this pr https://github.com/RedoxEngine/redox-hl7-v2/pull/2 mentions a bug where `parseInt` is not done properly here. maybe this is creating further issues.

**Comment by x-aaron-moore** [comment]
> This is awesome @blendersfun - I updated to code of conduct link, and appreciate you writing tests. Helps me wrap my head around what you're doing. Especially like the new schema test.
> 
> Overall, I'm ready to approve - but wondered about two things:
> 
> 1. `parseInt(element.maxOccurs === 1) ` that @sdhakal-xealth points out we have another PR for - what are your thoughts on this?

Yeah! I have been meaning to circle back and incorporate @sdhakal-xealth's feedback, but have been focused on other things. Will address this today.

> 2. We don't currently have npm publishing setup for this repo - would you be referencing the git commit in package.json or should I set up a GitHub action?

Yeah, so far we're referencing the git commit in package.json. But yes, it would be much easier to consume if a new npm version could get published.

**Comment by x-aaron-moore** [comment]
I did some more testing and found that the fix to that bug could result in data getting stripped out. Essentially if the definition says that only one item was expected, but an array was found in the message, it would write the segment 3 letter code and then the rest of the segment would be empty.

What the desired behavior should be in that case is not 100% clear. I went with the assumption that:

- The schema definition should be respected.
- Data should be written if at all possible.

This resulted in the behavior of picking the first element in the array and truncating the rest when the definition expects a single segment but multiple are found. It also ensures that for at least the first segment in the unexpected array, the data will always be written.

What do you think? Does that seem acceptable? It is a weird edge case.

**Comment by x-aaron-moore** [comment]
Just wanted to check in about this PR, @nwhatt, do you think my alterations are suitable for the parseInt issue? Am happy to incorporate feedback but was hoping to get this merged in at some point.

**Review by nwhatt — APPROVED** [review]

**Comment by nwhatt** [comment]
Hey @x-aaron-moore sorry about the delay - your explanation makes sense to me! Approved and merged!

**Comment by x-aaron-moore** [comment]
> Hey @x-aaron-moore sorry about the delay - your explanation makes sense to me! Approved and merged!

No worries! Thanks!

**Pull request merged**